### ISSUE TITLE
[UX] Revamp the maintenance mode page

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#2583]
I usually only ever visit this page if my site has been left in maintenance mode accidentally, and I need to disable that setting. In that case, I have clicked a link that says "Go online" and I expect by clicking that link, I will go online. Instead, I am taken to a form, where I usually blindly hit "Save configuration" since I can't tell that I need to UNcheck the box for `Put site into maintenance mode`.

A checkbox with a label that's an action is not the right tool here. The says "Put..." and that's not something my brain things can be toggled. An action has a yes/no answer, not an on/off value. 

What if the label were simply `Maintenance mode` and the form element was a set of radio buttons: `On` or `Off`?

Also, we should do a better job of letting people know the site is in maintenance mode:
1) how about a message at the top of this page saying my site is in maintenance mode
2) how about a message in the status report, saying my site is in maintenance mode

**Comment by klonos** [comment]
Related: #2529

**Comment by klonos** [comment]
Not sure if swapping the checkbox with a radio would improve things that much. I agree though with the whole action text for checkbox label being confusing. In fact @jenlampton this issue feels like a déjà vu from back in 2014: https://github.com/backdrop/backdrop-issues/issues/513#issuecomment-68236676

>...also, having to first check/uncheck a chechbox and then hit "save configuration" once in the maintenance page feels unnatural to me (going on/offline is an action and not a site config the way I see it). Can we replace the whole thing with a "Go online" / "Go offline" button depending on the site status? Once again, if there is fear of accidentally putting the site on/offline, we could have a confirmation dialog in place.

Here's what I think would be ideal:

1. Get rid of the checkbox altogether.
2. Add a toggle button with its label being either "Put the site to maintenance mode" or "Take the site off maintenance mode" ...or the simple, less verbose "Go offline"/"Go online".
3. The main purpose of the checkbox is to avoid accidental clicks taking the site off/online by requiring a second click on the form submit button. Since removing it would re-introduce that risk, we should have a confirmation dialog pop-up (#769, #770) upon pressing the toggle submit button.
4. Have the "Go online" link in the "site in maintenance mode" message work in a similar way, without redirecting the user to `/admin/config/development/maintenance` (#513)

**Comment by jenlampton** [comment]
> ..or the simple, less verbose "Go offline"/"Go online".

I don't think we should use the word "Online" as that's inaccurate. The site is still online when it's in maintenance mode!

> Add a toggle button 

I was thinking about a button too, but I wondered if that was the right UI pattern to use here. This isn't something like clearing a cache - there is actually a state that's being changed, and it would be useful if the UI displayed that clearly. I feel like button text (especially if it's verbose) won't be a clear enough indicator.

**Comment by klonos** [comment]
Although I agree that online/offline is not accurate, I think that it makes more sense to beginners and keeps the labels short and tidy. The opposite of "in maintenance mode" is "not in maintenance mode" and when displaying the site status, people that are speed-reading might miss the "not". Anyways, I'm not that rigorous about it and we can go either way since that is not the important part. Here's how that page could look in both states with the checkbox turned to a button:

![backdrop-issue2583-site_online](https://cloud.githubusercontent.com/assets/2423362/24063361/70975a46-0bb3-11e7-98f3-eeb8b87cbcec.png)

![backdrop-issue2583-site_offline](https://cloud.githubusercontent.com/assets/2423362/24063373/79872d52-0bb3-11e7-8aa6-35dc74381ba3.png)

1. There is message showing in that page at all times in order to better convey the site status with both text and color.
2. A better help text for "why" people would wanna place their site in maintenance mode is right under the action button.
3. The respective permission is embedded in the form to reduce pogo-sticking around admin pages.
4. Help text for the text message and the permission is moved between these two elements.

**Comment by jenlampton** [comment]
I still think it would be better if we displayed the actual state of the setting instead of using a button.

**Comment by klonos** [comment]
...by the way, the permission must have been changed to "Use the site in maintenance mode" while the message still refers to it as "Access site in maintenance mode".

**Comment by klonos** [comment]
>Are there any other UIs in core where we use a button to change the state of a setting?

The closest that I can think of is the way we add/remove things from the installer queue, but that is not saving config.

**Comment by klonos** [comment]
...then there's the "enable"/"disable"/"set default" for then themes, but these are links and not buttons.

**Comment by klonos** [comment]
So, @jenlampton were you thinking something along these lines?:

![backdrop-issue2583-radios_instead_of_checkbox](https://cloud.githubusercontent.com/assets/2423362/24064609/9154a9b8-0bb9-11e7-8306-0d6ec9299161.png)

**Comment by jenlampton** [comment]
"Site status" is also ambiguous.  Here's what I said:

> What if the label were simply `Maintenance mode` and the form element was a set of radio buttons: `On` or `Off`?

**Comment by klonos** [comment]
Yes, it does seem nice and clear, sorry I missed that. Should we then go with that + embed the permission at the bottom of that form?

**Comment by klonos** [comment]
...OK, I took a first stab at this. Here's how it looks:

<img width="1177" alt="screen shot 2018-10-13 at 5 06 23 pm" src="https://user-images.githubusercontent.com/2423362/46902095-cbc35980-cf0a-11e8-9d35-33596802deee.png">

Changes:

1. Changed the "Put site into maintenance mode" checkbox to a set of on/off "Maintenance mode" radios.
2. Added simplified/short descriptions to the radio options, and moved the previous long description to two separate messages (also see next point 3); one warning which includes what felt to be, ...well the warning bit + another info message, which includes what feels to be information about how things work.
3. The "The site is now in maintenance mode. ..." message is triggered upon rendering the form, rather than upon from submit. This allows the message to be shown when the maintenance mode settings page is visited, and serves as an additional way to communicate the current status (in case this is not immediately apparent from the state of the on/off radio).
4. There was a small bug with the "The site is no longer in maintenance mode." message popping up when it shouldn't, so I have slightly changed the logic to be more "defensive" and rely on `state_get('maintenance_mode', FALSE)` rather than the value of the radio on the form.
5. Added a `backdrop_get_messages();` in the submit handler to ensure that previous messages are not "bleeding" into the next page refresh (after consecutive switching between on/of of the maintenance mode).
6. Added a `cache_clear_all();` at the end of the submission handler, because in many cases while I was testing on my local, I was still able to access what seemed to be a cached version of the site, so I could not really tell if the switch from the checkbox to radios was working or not. We can keep that in or remove.
7. Permissions fieldset added at the bottom of the form.

Now, I might need some help with point 7, if I don't figure out why the permissions checkboxes are not mirrored between the maintenance mode page and the main permissions page.

**Comment by klonos** [comment]
...by the way, I noticed another bug that you might also notice in the PR sandbox, when clearing caches while in the maintenance mode settings page, while switching repeatedly between on/off. Turns out that there is a very small fix for that annoying bug, which applies to the current version of Backdrop, and is completely unrelated to the code changes in my PR here. So I have filed a separate ticket and PR for it here: #3313

**Comment by docwilmot** [comment]
👍

**Comment by klonos** [comment]
Thanks for the feedback @docwilmot 👍 ...

- So, the reason why I have split the message to a warning and an info is that the text was too long, and people tend to skip reading long messages. So, I have emphasised the "The site is currently in maintenance mode." bit in the warning + added the shortest explanation as to why this warrants a warning (the "Visitors will only see the maintenance mode message." bit). The rest felt like nice-to-know or help text. Plus, see point below...
- Yeah, I have left that "permission" link pointing to the main permissions page just in case we decided to show the info message in other admin pages besides the maintenance mode page ...i.e. move it here: https://github.com/backdrop/backdrop/blob/1.x/core/includes/menu.inc#L4038. I have now made it a link with an anchor to the id of the permissions fieldset wrapper (this is helpful in narrow screens, where the permissions fieldset might be outside the user viewport). Please let me know if you think that it'd be best to remove entirely (unless we decide to move the info to menu.inc).
- Good catch on the permission name! Fixed.

**Comment by klonos** [comment]
...so I was trying to troubleshoot the problem with the permission checkboxes. I did find a small issue [which was an oversight](https://github.com/backdrop/backdrop/pull/2320/commits/62a84aab49ad0068dce71c9537114fc306527d22), which I have fixed and pushed to my PR. I then continued troubleshooting, and have added `dpm`s before after the `foreach` [here](https://github.com/backdrop/backdrop/pull/2320/files#diff-2a3910dbb693d734a6867b68d5c83656R2120), like so:

```
dpm($permisson, "before");
    foreach ($roles as $role_name => $role) {
dpm($permission, "after");
```

...and I cannot figure out why the `$permisson` variable gets emptied when entering the foreach loop:

<img width="488" alt="screen shot 2018-10-14 at 8 49 57 am" src="https://user-images.githubusercontent.com/2423362/46910308-2eaa0480-cf8e-11e8-94c7-b649c1d246db.png">

**Comment by amoutiers** [comment]
Hi @klonos, I've just checkout your PR and find the problem, it's just a typo
https://github.com/backdrop/backdrop/pull/2320

**Comment by klonos** [comment]
Hello @amoutiers and thank you for the review 👍 ...can you please provide more information about the typo you have discovered? Thanks

**Comment by amoutiers** [comment]
@klonos, it’s on the PR page ;-)

**Comment by amoutiers** [comment]
On the PR comment page https://github.com/backdrop/backdrop/pull/2320
I’ve started a review that suggest a change for the typo who lead to your problem

**Comment by klonos** [comment]
Thanks @amoutiers ...good catch 👍 Fixed.

**Comment by klonos** [comment]
Hey @docwilmot ...

> - dont like that there are two persistent message types (blue and orange)

How about this then @docwilmot?:

<img width="853" alt="screen shot 2018-12-11 at 8 39 15 am" src="https://user-images.githubusercontent.com/2423362/49782819-5c0df700-fd20-11e8-96b5-97643239016d.png">

- added a "(as configured below)" bit in the warning.
- reordered the on/off radios, so that the maintenance message comes right after the "On". This way the "Visitors will only see the maintenance mode message." help text is closer to the message text box.
- moved the info message inline in the form, right above the permissions section (and removed the link)

**Comment by klonos** [comment]
...or the message can be `info` while not in maintenance mode:

<img width="871" alt="screen shot 2018-12-11 at 9 05 31 am" src="https://user-images.githubusercontent.com/2423362/49783911-24a14980-fd24-11e8-9484-131fbae77768.png">

...and a `warning` when maintenance mode is on:

<img width="863" alt="screen shot 2018-12-11 at 9 05 57 am" src="https://user-images.githubusercontent.com/2423362/49783914-28cd6700-fd24-11e8-8a2d-697aa8270413.png">

**Comment by klonos** [comment]
...the toggle button does not save the other form settings (maintenance mode message and permissions) in its current implementation. I would like to get some feedback first, and if people think that this is what should happen, then I think that this would be a simple matter of chaining `system_site_maintenance_mode_submit()` from `system_maintenance_mode_toggle_submit()`.

**Comment by olafgrabienski** [comment]
> the toggle button does not save the other form settings in its current implementation

The other settings shouldn't be saved, in my opinion, I wouldn't expect it. See also the other examples (Clear caches, Invalidate search index) where other configuration settings aren't saved when only using the specific action button.

**Comment by klonos** [comment]
> The other settings shouldn't be saved, in my opinion, I wouldn't expect it. ...

I am leaning towards that myself @olafgrabienski (hence implemented like so in the PR). My only concern/dilemma is this:

The "Maintenance mode message" textfield feels like better to be grouped in the same fieldset as the toggle button, but being in the same fieldset makes me expect that the button closest to that textfield and within the same fieldset would also save any change in the message text.

...in combination with this (which is actually a data loss issue):

If the user updates the "Maintenance mode message" text and hits the toggle button, then they lose their changes, and have to re-edit and submit the form to actually save them.

On the other hand, if the user had edited the message text before clicking the toggle button, is there any logical reason why they would intentionally want their changes not to be saved? So what would be the harm if we also saved the form here in one go? I mean, would you be upset if by clicking the toggle button also saved the text (and perhaps also the permissions)?

One last thing to consider is the use case where a person wants to turn the maintenance mode on/off, and **at the same time** make changes to the message text and/or the permissions. With the toggle being a checkbox/radio, they would make the changes and submit the form once. With the current implementation of the toggle button, they would need to first toggle maintenance, then edit and submit the form, which feels like a two-step process instead of a single-step one. Not necessarily a bad thing; just putting it out there for consideration.

Anyway, the last thing I want to do is to derail this and miss the 1.12.0 release, because I really find the toggle button to be a much more natural UX than the checkbox or the radio buttons. At the same time I do not want this to be a weird UX for others, so I would like to hear what others think. I personally think that all solutions have their quirks, so at the end we'll need to decide which one is the least awkward 😄

**Comment by olafgrabienski** [comment]
> would you be upset if clicking the toggle button also saved the text?

Most of the times not, even if I can imagine some situations (which are not very likely to happen). It's however more a consistency question: If we change the behavior to saving the whole form, I guess I'd generally feel insecure how these sort of buttons are supposed to work.

We could also consider to change the text to "Turn maintenance mode on and Save configuration", but that's quite long.

**Comment by klonos** [comment]
True re consistency @olafgrabienski ...as I said, it's the reason why I have not implemented saving configuration as well in the current PR.

> We could also consider to change the text to "Turn maintenance mode on and Save configuration", but that's quite long.

Yeah, I've thought of that too ...and I dismissed it once I have tried it out on my local:

<img width="921" alt="screen shot 2018-12-17 at 11 34 18 pm" src="https://user-images.githubusercontent.com/2423362/50117175-e391c380-0254-11e9-953b-097d7cde2aa3.png">

...the button was so long, that part of it would get rendered outside the viewport:

![screencapture-trusted-hosts-local-admin-config-development-maintenance-2018-12-17-23_35_43](https://user-images.githubusercontent.com/2423362/50117624-315afb80-0256-11e9-82e8-c22c45f8f55d.png)

Besides, in that case, we'd fall into the same UI pattern used when adding content types and views, where the "dual action" submit buttons are positioned at the end of the form:

<img width="578" alt="screen shot 2018-12-17 at 11 44 27 pm" src="https://user-images.githubusercontent.com/2423362/50117536-f35dd780-0255-11e9-99c9-d7d408c7b24c.png">

<img width="516" alt="screen shot 2018-12-17 at 11 40 07 pm" src="https://user-images.githubusercontent.com/2423362/50117541-f5279b00-0255-11e9-878e-9b8196191711.png">

...and we'd most likely need to do something like this instead:

<img width="912" alt="screen shot 2018-12-17 at 11 54 13 pm" src="https://user-images.githubusercontent.com/2423362/50118042-28b6f500-0257-11e9-8f34-d11f9a299b41.png">

**Comment by jenlampton** [comment]
@klonos I don't like the button at the bottom of this page, it's the most important setting here, and can easily be missed now. Most people won't read these secondary buttons, as they've been trained to blindly click the blue ones in most cases. I'm going to review the first PR with the radios, perhaps we can revisit the button idea for 1.13?

**Comment by jenlampton** [comment]
I've filed a few notes on the PR, but it looks like it has conflicts. I think this is an improvement - especially with the permissions right here - but I'm conflicted about the info message that's not a message. That should really be `#help` instead.

**Comment by jenlampton** [comment]
removing the 1.12 milestone, but adding the label for milestone review.

**Comment by klonos** [comment]
I have tried to fix the failing tests in the PR with the 2 radio options. I got one fixed easily (was a string change), but the one left eludes me 😞😓

Pinging @jenlampton ...I would like to get this in for 1.13 if possible. So:

- go with the radio options, or do you think a single toggle button (like the "Clear caches" and "Invalidate search index")?
- change the info message to `#help`
- ???

What else to move this along?

**Comment by jenlampton** [comment]
> go with the radio options

Let's do this first. The other two examples are for actions that are performed on a separate form, in this case the setting is what the form is for. We can debate the change later if people prefer that approach, but for now...  baby steps :)

**Comment by philsward** [comment]
So... what about moving the entire "maintenance mode" functionality to the admin menu with a "Live" or "Maintenance" (Not Live?) indicator (Swap words for Icons and colors?) and clicking on the indicator would switch between the two modes. Remove the checkbox that is on the maintenance admin page and just leave the custom textbox for a custom message. Going in and out of maintenance mode shouldn't take 3 clicks and it would make it that much easier for the end-user who isn't a site builder to move their site from live to maintenance.

I think it would simplify the discussion going on here to stop using the maintenance page as the enabler/disabler and it would remove the need for the status message that interferes with theme styling unless you yank it out of the inspector each refresh.

I think I've put my various combined sites into maintenance mode maybe 5-10 times in the last 15 years. I might use it more if it were more readily available. A good candidate for Telemetry?
